doc/crypto/api/ops/key-agreement.rst

```diff
@@ -332,6 +338,264 @@ Combining key agreement and key derivation
+
+        .. code-block:: xref
+
+            psa_key_agreement_iop_t operation;
+            operation = psa_key_agreement_iop_init();
+
+    This is an implementation-defined type.
+    Applications that make assumptions about the content of this object will result in implementation-specific behavior, and are non-portable.
+
+.. macro:: PSA_KEY_AGREEMENT_IOP_INIT
+    :definition: /* implementation-defined value */
+
+    .. summary::
+        This macro evaluates to an initializer for an interruptible key agreement operation object of type `psa_key_agreement_iop_t`.
+
+.. function:: psa_key_agreement_iop_init
+
+    .. summary::
+        Return an initial value for an interruptible key agreement operation object.
+
+    .. return:: psa_key_agreement_iop_t
+
+.. function:: psa_key_agreement_iop_get_num_ops
+
+    .. summary::
+        Get the number of *ops* that an interruptible key agreement operation has taken so far.
+
+    .. param:: psa_key_agreement_iop_t * operation
+        The interruptible key agreement operation to inspect.
+
+    .. return:: uint32_t
+        Number of *ops* that the operation has taken so far.
+
+    After the interruptible operation has completed, the returned value is the number of *ops* required for the entire operation.
+    The value is reset to zero by a call to either `psa_key_agreement_iop_setup()` or `psa_key_agreement_iop_abort()`.
+
+    This function can be used to tune the value passed to `psa_iop_set_max_ops()`.
+
+    The value is undefined if the operation object has not been initialized.
+
+.. function:: psa_key_agreement_iop_setup
+
+    .. summary::
+        Start an interruptible operation to perform a key agreement.
+
+    .. param:: psa_key_agreement_iop_t * operation
+        The interruptible key agreement operation to set up.
+        It must have been initialized as per the documentation for `psa_key_agreement_iop_t`, and be inactive.
+    .. param:: psa_key_id_t private_key
+        Identifier of the private key to use.
+        It must permit the usage `PSA_KEY_USAGE_DERIVE`.
+    .. param:: const uint8_t * peer_key
+        Public key of the peer.
+        The peer key data is parsed with the type :code:`PSA_KEY_TYPE_PUBLIC_KEY_OF_KEY_PAIR(type)` where ``type`` is the type of ``private_key``, and with the same bit-size as ``private_key``.
+        The peer key must be in the format that `psa_import_key()` accepts for this public key type.
+        These formats are described with the public key type in :secref:`key-types`.
+    .. param:: size_t peer_key_length
+        Size of ``peer_key`` in bytes.
+    .. param:: psa_algorithm_t alg
+        The standalone key agreement algorithm to compute: a value of type `psa_algorithm_t` such that :code:`PSA_ALG_IS_STANDALONE_KEY_AGREEMENT(alg)` is true.
+    .. param:: const psa_key_attributes_t * attributes
+        The attributes for the new key.
+        This function uses the attributes as follows:
```

Comment:
The bullet points below actually mostly explain how the resulting key is constructed rather than how the `attributes` parameter is used. I find this presentation confusing. In particular, the key size is always the same regardless of the attributes. Separately, the key size in the attributes must be either correct or 0.

Reply:
This text is identical to that for `psa_key_agreement()`, which we discussed at some length as part of https://github.com/ARM-software/psa-api/pull/101 (specifically commit https://github.com/ARM-software/psa-api/pull/101/commits/5c48afe702061b077754cf60156b4b1972713742).

Is there a good reason to do something different here, or to change what we have done in the existing APIs?

Reply:
Sorry, I missed that earlier. My objection isn't specific to this function. The explanations of how the attributes are used and how the new key is constructed are confusing.

Reply:
Proposal for improvement raised as #205. Will feed results of that PR into this one.

Reply:
Clarifications in the updated PR #205 have been copied into this PR